src/full/Agda/Interaction/CommandLine.hs

```diff
@@ -34,36 +34,54 @@ import Agda.TypeChecking.Pretty ( PrettyTCM(prettyTCM) )
 import Agda.TypeChecking.Substitute
 import Agda.TypeChecking.Warnings (runPM)
 
-import Agda.Utils.Monad
+import Agda.Utils.FileName (absolute, AbsolutePath)
+import Agda.Utils.Maybe (caseMaybeM)
 import Agda.Utils.Pretty
 
 import Agda.Utils.Impossible
 
+data ReplEnv = ReplEnv
+    { replSetupAction     :: TCM ()
+    , replTypeCheckAction :: AbsolutePath -> TCM (Maybe Interface)
+    }
+
+data ReplState = ReplState
+    { currentFile :: Maybe AbsolutePath
+    }
+
+type ReplM = ReaderT ReplEnv (StateT ReplState IM)
```

Comment:
Could this be a `newtype .... deriving (... MonadReader ReplEnv, MonadState ReplState...)`?

Reply:
I think I tried that and gave up in frustration. But I might have missed something.

Reply:
I must have missed something. No problems there.

Reply:
That's pushed